src/diffusers/models/autoencoders/dc_ae.py

```diff
@@ -0,0 +1,877 @@
+        ]
+    )
+    if shortcut is None:
+        pass
+    elif shortcut == "averaging":
+        shortcut_block = PixelUnshuffleChannelAveragingDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=1
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for encoder project out")
+    return block
+
+
+def build_decoder_project_in_block(in_channels: int, out_channels: int, shortcut: Optional[str]):
+    block = ConvLayer(
+        in_channels=in_channels,
+        out_channels=out_channels,
+        kernel_size=3,
+        stride=1,
+        use_bias=True,
+        norm=None,
+        act_func=None,
+    )
+    if shortcut is None:
+        pass
+    elif shortcut == "duplicating":
+        shortcut_block = ChannelDuplicatingPixelUnshuffleUpsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=1
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for decoder project in")
+    return block
+
+
+def build_decoder_project_out_block(
+    in_channels: int, out_channels: int, factor: int, upsample_block_type: str, norm: Optional[str], act: Optional[str]
+):
+    layers: list[nn.Module] = [
+        build_norm(norm, in_channels),
+        get_activation(act) if act is not None else None,
+    ]
+    if factor == 1:
+        layers.append(
+            ConvLayer(
+                in_channels=in_channels,
+                out_channels=out_channels,
+                kernel_size=3,
+                stride=1,
+                use_bias=True,
+                norm=None,
+                act_func=None,
+            )
+        )
+    elif factor == 2:
+        layers.append(
+            build_upsample_block(
+                block_type=upsample_block_type, in_channels=in_channels, out_channels=out_channels, shortcut=None
+            )
+        )
+    else:
+        raise ValueError(f"upsample factor {factor} is not supported for decoder project out")
+    return OpSequential(layers)
+
+
+class Encoder(nn.Module):
+    def __init__(
+        self, 
+        in_channels: int,
+        latent_channels: int,
+        width_list: list[int] = [128, 256, 512, 512, 1024, 1024],
+        depth_list: list[int] = [2, 2, 2, 2, 2, 2],
+        block_type: str | list[str] = "ResBlock",
+        norm: str = "rms2d",
+        act: str = "silu",
+        downsample_block_type: str = "ConvPixelUnshuffle",
+        downsample_match_channel: bool = True,
+        downsample_shortcut: Optional[str] = "averaging",
+        out_norm: Optional[str] = None,
+        out_act: Optional[str] = None,
+        out_shortcut: Optional[str] = "averaging",
+        double_latent: bool = False,
+    ):
+        super().__init__()
+        num_stages = len(width_list)
+        self.num_stages = num_stages
+        assert len(depth_list) == num_stages
```

Comment:
We avoid doing assertions and instead, just raise `ValueError`s. In this case, we could have a function like `validate_config()` and raise errors from there.

Reply:
Got it.